core/include/detray/tools/constrained_step.hpp

```diff
@@ -0,0 +1,117 @@
+/** Detray library, part of the ACTS project (R&D line)
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// detray definitions
+#include <climits>
+#include <type_traits>
+
+#include "detray/definitions/qualifiers.hpp"
+
+namespace detray {
+
+// Add constraints for steppers
+namespace step {
+
+/// Direction in which the integration is performed
+enum direction : int {
+    e_forward = 1,
+    e_unknown = std::numeric_limits<int>::max(),
+    e_backward = -1,
+};
+
+/// the types of constraints
+/// from accuracy - this can vary up and down given a good step estimator
+/// from actor    - this would be a typical navigation step
+/// from aborter  - this would be a target condition
+/// from user     - this is user given for what reason ever
+enum constraint : std::size_t {
```

Comment:
Please don't mind my ignorance. Could you explain which constraints we have to use for which cases? It seems `steppers` are only using `e_all`.

Reply:
It is not possible to set a constraint for ```e_all```. This is only used to retrieve the strictest constraint among all constraints. Only the strictest of those constraints is needed in the stepper.

Reply:
That's what I mean - If we only consider strictest of those constraints, only one constraint wouldn't be enough?

Reply:
Okay makes sense